src/Network/Wai/Log/Internal.hs

```diff
@@ -1,39 +1,52 @@
 {-# LANGUAGE RecordWildCards #-}
 module Network.Wai.Log.Internal where
 
-import Control.Monad (when)
-import Data.Aeson.Types (Value, object, emptyObject)
+import Data.Aeson.Types (Value, object)
+import Data.ByteString.Builder (Builder, toLazyByteString)
 import Data.Text (Text)
 import Data.Time.Clock (UTCTime, diffUTCTime, getCurrentTime)
+import Data.UUID (UUID)
+import Data.UUID.V4 (nextRandom)
 import Log (LogLevel)
-import Network.Wai (Middleware)
+import Network.HTTP.Types.Status (statusIsClientError, statusIsServerError)
+import Network.Wai (Application, responseToStream)
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.Text.Encoding as T
 
 import Network.Wai.Log.Options (Options(..), ResponseTime(..))
 
 -- | This type matches the one returned by 'getLoggerIO'
 type LoggerIO = UTCTime -> LogLevel -> Text -> Value -> IO ()
 
--- | Create a logging 'Middleware' given a 'LoggerIO' logging function and 'Options'
-logRequestsWith :: LoggerIO -> Options -> Middleware
-logRequestsWith loggerIO Options{..} app req respond = do
-  logIO "Request received" . object . logRequest $ req
+logRequestsWith :: LoggerIO -> Options -> (UUID -> Application) -> Application
+logRequestsWith loggerIO Options{..} mkApp req respond = do
+  uuid <- nextRandom
+  logIO "Request received" $ logRequest uuid req
   tStart <- getCurrentTime
-  app req $ \resp -> do
+  mkApp uuid req $ \resp -> do
     tEnd <- getCurrentTime
-    when logSendingResponse $
-         logIO_ "Sending response"
     r <- respond resp
     tFull <- getCurrentTime
     let processing = diffUTCTime tEnd  tStart
         full       = diffUTCTime tFull tStart
         times      = ResponseTime{..}
-    logIO "Request complete" . object $ logResponse req resp times
+
+    let (status, _headers, bodyToIO) = responseToStream resp
+    _ <- if statusIsClientError status || statusIsServerError status
+      then
+        bodyToIO $ \streamingBody ->
+          let builderToText :: Builder -> Text
+              builderToText b = T.decodeUtf8 $ BSL.toStrict $ toLazyByteString b
+              logWithBuilder :: Builder -> IO ()
+              logWithBuilder b =
+                let body = Just $ builderToText b
+                in logIO "Request complete" $ logResponse uuid req resp body times
+          in streamingBody logWithBuilder (return ())
```

Comment:
StreamingBody is a function that takes two arguments: 
```
the first parameter provides a means of sending another chunk of data, 
and the second parameter provides a means of flushing the data to the client.
```
https://hackage.haskell.org/package/wai-3.2.2.1/docs/Network-Wai.html#t:StreamingBody

I'm not sure about the second one. 

Also, if body consists of more than one chunks, then the `logIO "Request complete"` will be called multiple times as I understand it. I don't know how one could work around this thing. It doesn't happen in the practice when I tried, because our response body is small, but it could happen potentially I assume.

Reply:
Hm, maybe this only applies to https://en.wikipedia.org/wiki/Chunked_transfer_encoding so then we're fine. I've tried with a long body message (10000 words) and it logged it in 1 message.

Reply:
Ok, I will need to dig into this a bit before we merge, but just quickly looking at `T.decodeUtf8 $ BSL.toStrict $ toLazyByteString b` we are making some incorrect assumptions.

We should not assume anything about the response!
Currently we have assumed that:
* it has already been constructed (`toLazyByteString`)
* it will fit in memory (`BSL.toStrict`)
* it is UTF8 (`decodeUtf8`)

While the last assumption is most flagrant problem, the others shouldn't be ignored either.

Reply:
So, I've moved that code to `eid-service` and `wai-log` only take a function `Status -> Builder -> Value`, so it could build a loggable value on the basis of this information. How do you like it?